server/src/main/java/de/uftos/repositories/ucdl/parser/DefinitionParser.java

```diff
@@ -0,0 +1,691 @@
+          return new OperatorDto(UcdlToken.SIZE, set);
+        } else {
+          return new ValueDto<>(UcdlToken.NUMBER, Integer.parseInt(
+              ((SimpleNode) root).jjtGetFirstToken().image));
+        }
+      }
+      case "SET" -> {
+        AbstractSyntaxTreeDto setName = buildAst(root.jjtGetChild(0).jjtGetChild(0), parameters);
+
+        ResourceType setType;
+
+        if (setName.getToken() == UcdlToken.NUMBER_SET) {
+          setType = ResourceType.NUMBER;
+        } else {
+          setType = ((ElementDto) setName).type();
+        }
+
+        List<AbstractSyntaxTreeDto> modifiers = new ArrayList<>();
+
+        Node modifierList = root.jjtGetChild(1);
+        while (modifierList.jjtGetNumChildren() > 0) {
+          if (modifierList.jjtGetNumChildren() == 2) {
+            SimpleNode flatmap = (SimpleNode) modifierList.jjtGetChild(0);
+            String attribute = flatmap.jjtGetFirstToken().image;
+            setType = getNextResourceType(setType, attribute);
+            modifiers.add(new ValueDto<>(UcdlToken.ATTRIBUTE, attribute));
+            modifierList = modifierList.jjtGetChild(1);
+
+          } else if (modifierList.jjtGetNumChildren() == 3) {
+            ResourceType thisType = parameters.get("this");
+            parameters.replace("this", setType);
+
+            List<AbstractSyntaxTreeDto> filters = new ArrayList<>();
+            filters.add(buildAst(modifierList.jjtGetChild(0), parameters));
+
+            Node filterList = modifierList.jjtGetChild(1);
+            while (filterList.jjtGetNumChildren() > 0) {
+              filters.add(buildAst(filterList.jjtGetChild(0), parameters));
+              filterList = filterList.jjtGetChild(1);
+            }
+
+            for (AbstractSyntaxTreeDto filter : filters) {
+              if ((filter.getToken() == UcdlToken.NUMBER_SET && setType != ResourceType.NUMBER)
+                  || (filter.getToken() == UcdlToken.RESOURCE_SET
+                  && ((SetDto) filter).type() != setType)) {
+                throw new ParseException("Sets can only be filtered using sets of the same type!");
+              }
+            }
+
+            modifiers.add(new OperatorDto(UcdlToken.FILTER, filters));
+            modifierList = modifierList.jjtGetChild(2);
+            parameters.replace("this", thisType);
+          }
+        }
+
+        return new SetDto(UcdlToken.RESOURCE_SET, setType, setName, modifiers);
+      }
+      case "NUMBERSET" -> {
+        List<Integer> values = new ArrayList<>();
+
+        values.add(Integer.parseInt(((SimpleNode) root).jjtGetFirstToken().next.image));
+
+        SimpleNode numberList = (SimpleNode) root.jjtGetChild(0);
+        while (numberList.jjtGetNumChildren() > 0) {
+          values.add(Integer.parseInt(numberList.jjtGetFirstToken().next.image));
+          numberList = (SimpleNode) numberList.jjtGetChild(0);
+        }
+        return new ValueDto<>(UcdlToken.NUMBER_SET, values.toArray(new Integer[0]));
```

Comment:
That you can't write `values.toArray<Integer>()` still baffles me. Average java moment lol

Reply:
You can also just use  `values.toArray()`, however this would just return an `Object[]`.

Reply:
Exactly. I'm annoyed that you can't just type it using generics

Reply:
Same.
Generics would be cleaner.